File: Data Structures & Algorithms/depth-of-binary-tree/submission-1.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn max_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {

        let mut queue = VecDeque::new();

        if let Some(r) = root{
            queue.push_back(r);
        }

        let mut level = 0;

        while !queue.is_empty(){
            let size = queue.len();

            level += 1;

            for _ in 0..size{
                let node = queue.pop_front().unwrap();
                let node = node.borrow();

                if let Some(ref left) = node.left{
                    queue.push_back(left.clone());
                }
            
                if let Some(ref right) = node.right{
                    queue.push_back(right.clone());
                }
            
            }

        }
        
        level

    }
}

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[][]}
     */
    levelOrder(root) {
        if(!root) return []
        const queue = [root]

        const result = []

        while(queue.length > 0){
            let currLevelSize = queue.length
            let levelElements = []

            for(let i = 0; i < currLevelSize; i ++){
                const node = queue.shift();

                levelElements.push(node.val)
                if(node.left){queue.push(node.left)}
                if(node.right){queue.push(node.right)}
            }

            result.push(levelElements)
        }


        return result;
    }
}
